### DISCUSSION TITLE
Is there a way to save all/a single freetext bookmark to PNG?

### DISCUSSION THREAD

**Discussion by zetashift** [ahrm/sioyek#1249] [Q&A] [answered]
I would like to save bookmarks made as a freetext bookmark, `:add_freetext_bookmark` and then bookmark text: `#r`, in an attachments folder as a JPG or PNG(any of these will suffice).

I use the freetext bookmark command as an area highlight/annotator and would like to reference these in images in my notes.
Does the extension API allow this?

**Comment by ahrm** [comment, marked as answer]
This was not possible before, but I added some new javascript api commands which makes it possible now. You can do something like this:

`screenshot.js`:
```
let annots = sioyek_api.get_json_annotations();
let state = sioyek_api.get_json_state();
let checksum = state["document_checksum"]; // this is only used to differentiate the file names of different documents
let screenshots_path = "e:/sioyek_screenshots/"

var index = 0;
for (let bookmark of annots["bookmarks"]){
    index++;

    if (bookmark["description"].startswith("#r")){
        let bookmark_abs_rect = {
            "x0": bookmark["begin_x"],
            "x1": bookmark["end_x"],
            "y0": bookmark["begin_y"],
            "y1": bookmark["end_y"]
        };

        sioyek_api.goto_offset(0, bookmark["begin_y"]);
        let bookmark_window_rect =  sioyek_api.absolute_to_window_rect_json(bookmark_abs_rect);
        sioyek_api.screenshot_js(screenshots_path + checksum +  "_screenshot" + index + ".png", bookmark_window_rect);

    }
}
```
`prefs_user.config`:
```
new_js_command _save_screenshots screenshot.js
```

**Reply by zetashift** [reply]
Ah sweet! Thank you very much! Am I right in that I need to build from `development` for this API?

Also just because this snippet is so straightforward(props on that!), it seems to be possible to also export the text annotations from `sioyek.api.get_json_annotations()` to a file?

Could be cool to export those to a markdown file with a file link to the text annotation location

**Reply by ahrm** [reply]
Yes you need the development branch for that. Yes, you can use that to export the annotations but we already have an `export` command that exports the annotations into a json file.
